L26-serialization-traits-per-call.md

```diff
@@ -0,0 +1,43 @@
+Serialization Mechanism Customizable Per Call
+----
+* Author(s): ncteisen
+* Approver: vjpai
+* Status: Draft
+* Implemented in: n/a
+* Last updated: 02/28/18
+* Discussion at: n/a
+
+## Abstract
+
+This document proposes a change to [SerializationTraits](https://github.com/grpc/grpc/blob/master/include/grpcpp/impl/codegen/serialization_traits.h) in order to allow the serialization mechanism to be configurable per call. Most use cases will be unaffected, but this will make it much easier for user to inject custom serialization code into the gRPC pathway.
+
+## Background
+
+Currently, gRPC supports a mechanism for custom serialization. Users must implement the `grpc::SerializationTraits<Message, void>` class for their specific Message type. Forcing users to implement a class that exists in the grpc namespace is a form of internal API leakage. Additionally, this method does not give users the granularity to change the serialization system per-call.
+
```

Comment:
IMO, valuable background would be to give a comparison of how the other languages do this. I know that Go has customizable Codec per channel, not sure about Java.

Reply:
grpc-go is now customizable per-call.  And we have a registry to support multiple codecs in both clients and servers.  Check out this PR that implements what support we have; you may want to do something similar:

https://github.com/grpc/grpc-go/pull/1813

Usage docs are here: https://github.com/grpc/grpc-go/blob/master/Documentation/encoding.md (this also discusses our compressor support that works the same way).

Reply:
Thanks for the info! I had seen it as a DialOption but good to know that it is now a CallOption. This is valuable background feedback.

@ncteisen : Can you also retitle the gRFC itself to indicate that it's L26 and that it's C++ specific? Likewise with references to gRPC converting to gRPC C++

Reply:
I added info on Java and Go